File: frontend/src/types/lagrange.ts
export interface PointDTO {
  x: number;
  y: number;
}

export interface EvaluateRequest {
  func: string;
  points_type: 'equispaced' | 'custom';
  a: number;
  b: number;
  n: number;
  x_eval: number;
  custom_points?: PointDTO[];
}

export interface EvaluateResponse {
  x_eval: number;
  interpolated_value: number;
  derivative_approx: number;
  derivative_exact: number;
  absolute_error: number;
  relative_error: number;
  execution_time: number;
  num_points: number;
}

export interface ExperimentRequest {
  func: string;
  a: number;
  b: number;
  x_eval: number;
  n_values: number[];
  plot_n?: number;
}

export interface ExperimentResultDTO {
  n: number;
  result: number;
  exact_value: number;
  absolute_error: number;
  relative_error: number;
  execution_time: number;
  iterations: number;
}

export interface FunctionPlotPoint {
  x: number;
  f_x: number;
  p_x: number;
}

export interface ExperimentResponse {
  function_name: string;
  a: number;
  b: number;
  x_eval: number;
  results: ExperimentResultDTO[];
  function_plot_data: FunctionPlotPoint[];
  interpolation_points: PointDTO[];
}

export interface FunctionInfo {
  name: string;
  expression: string;
  label: string;
}

export interface FunctionsListResponse {
  functions: FunctionInfo[];
}

export interface BasisStep {
  i: number;
  basis_term: string;
  basis_simplified: string;
  contribution: string;
}

export interface StepsRequest {
  points: PointDTO[];
  x_eval: number;
}

export interface StepsResponse {
  points: PointDTO[];
  steps: BasisStep[];
  polynomial: string;
  derivative: string;
  evaluated: number;
  x_eval: number;
}

File: backend/app/domain/models/lagrange.py
from dataclasses import dataclass, field
from typing import Callable, List, Optional, Tuple


@dataclass(frozen=True)
class Point:
    x: float
    y: float


@dataclass
class LagrangeResult:
    x_eval: float
    interpolated_value: float
    derivative_approx: float
    derivative_exact: float
    absolute_error: float
    relative_error: float
    execution_time: float
    num_points: int


@dataclass
class ExperimentResult:
    n: int
    result: float
    exact_value: float
    absolute_error: float
    relative_error: float
    execution_time: float
    iterations: int


@dataclass
class ExperimentReport:
    function_name: str
    a: float
    b: float
    x_eval: float
    results: List[ExperimentResult] = field(default_factory=list)
    plot_data: List[Tuple[float, float, float]] = field(default_factory=list)
    interpolation_points: List[Point] = field(default_factory=list)


@dataclass
class EvaluatedFunction:
    name: str
    expression: str
    func: Optional[Callable[[float], float]]
    derivative: Optional[Callable[[float], float]]
